File: IP/3rd/sampleprogram/6-3-4-2.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;

int main()
{
  // 画像読み込み
  string file_src = "src.png";
  Mat img_src = imread(file_src, 1);

  int gamma = 1;

  // ウインドウ生成
  namedWindow("src", WINDOW_AUTOSIZE);
  namedWindow("dst", WINDOW_AUTOSIZE);
  // トラックバーの生成
  createTrackbar("gamma", "dst", &gamma, 10);
  // 入力画像を表示
  imshow("src", img_src);

  while(1) {
    // トラックバーの値を取得
    int gamma = getTrackbarPos("gamma", "dst") + 1;

    // ガンマ補正
    Mat img_dst;
    Mat lut(256, 1, CV_8U);
    for(int i = 0; i < 256; i++) {
      lut.data[i] = (unsigned char)(255.0 * pow(i / 255.0, 1.0 / gamma));
    }
    LUT(img_src, lut, img_dst);

    // 出力画像を表示
    imshow("dst", img_dst);

    // qキーで終了
    if(waitKey(1) == 'q') break;
  }
  return 0;
}

File: OpenCVによる画像処理入門/第3版/第2刷/sampleprogram/6-3-4-2.py
import cv2
import numpy as np

def nothing(x):
  pass

# 画像読み込み
file_src = 'src.png'
img_src = cv2.imread(file_src, cv2.IMREAD_COLOR)
# ウインドウ生成
cv2.namedWindow('src')
cv2.namedWindow('dst')
# トラックバーの生成
cv2.createTrackbar('gamma', 'dst', 1, 10, nothing)
# 入力画像を表示
cv2.imshow('src', img_src)

while True:
  # トラックバーの値を取得
  gamma = cv2.getTrackbarPos('gamma', 'dst') + 1.0

  # ガンマ補正
  Y = np.ones((256, 1), dtype = 'uint8') * 0
  for i in range(256):
    Y[i][0] = 255 * pow(float(i) / 255, 1.0 / gamma)
  img_dst = cv2.LUT(img_src, Y)

  # 出力画像を表示
  cv2.imshow('dst', img_dst)
  # qキーで終了
  key = cv2.waitKey(1)
  if key & 0xFF == ord('q'):
    break

cv2.destroyAllWindows()
